src/data/tree.lean

```diff
@@ -81,4 +81,64 @@ def map {β} (f : α → β) : tree α → tree β
 | nil := nil
 | (node a l r) := node (f a) (map l) (map r)
 
+/-- A non-nil `unit_tree`; useful when we want an arbitrary value other than `nil` -/
+abbreviation non_nil [inhabited α] : tree α := node default nil nil
+
+@[simp] lemma non_nil_ne [inhabited α] : non_nil ≠ (@nil α) := by trivial
+
+/-- The number of internal nodes (i.e. not including leaves) of a binary tree -/
+@[simp] def nodes : tree α → ℕ
+| nil := 0
+| (node _ a b) := a.nodes + b.nodes + 1
+
+/-- The number of leaves of a binary tree -/
+@[simp] def leaves : tree α → ℕ
+| nil := 1
+| (node _ a b) := a.leaves + b.leaves
+
+/-- The height - length of the longest path from the root - of a binary tree -/
+@[simp] def height : tree α → ℕ
+| nil := 0
+| (node _ a b) := max a.height b.height + 1
+
+lemma leaves_eq_nodes_succ (x : tree α) : x.leaves = x.nodes + 1 :=
+by { induction x; simp [*, nat.add_comm, nat.add_assoc, nat.add_left_comm], }
+
+lemma leaves_pos (x : tree α) : 0 < x.leaves :=
+by { rw leaves_eq_nodes_succ, exact x.nodes.zero_lt_succ, }
+
+lemma height_le_nodes : ∀ (x : tree α), x.height ≤ x.nodes
+| nil := le_refl _
+| (node _ a b) := nat.succ_le_succ
+    (max_le
+      (trans a.height_le_nodes $ a.nodes.le_add_right _)
+      (trans b.height_le_nodes $ b.nodes.le_add_left _))
+
+/-- The left child of the tree, or `nil` if the tree is `nil` -/
+@[simp] def left : tree α → tree α
+| nil := nil
+| (node _ l r) := l
+
+/-- The right child of the tree, or `nil` if the tree is `nil` -/
+@[simp] def right : tree α → tree α
+| nil := nil
+| (node _ l r) := r
+
+/-- Abbreviation for making a node with `unit` data -/
+@[pattern] abbreviation unode : tree unit → tree unit → tree unit :=
+tree.node ()
```

Comment:
Everything before that point isn't mentioned anywhere in the PR description (which was already outdated). Could you either move that material in a new PR (it looks quite self-contained) or mention it in the PR description?

Reply:
Mentioned it in the PR description.